rustls/src/dns_name.rs

```diff
@@ -0,0 +1,176 @@
+/// DNS name validation according to RFC1035, but with underscores allowed.
+use std::error::Error as StdError;
+use std::fmt;
+
+/// "Labels must be 63 characters or less."
+const MAX_LABEL_LENGTH: usize = 63;
+
+fn validate(input: &[u8]) -> Result<(), InvalidDnsNameError> {
+    let mut label_length = 0;
+    let mut previous_label_ch = b'-';
+
+    for ch in input {
+        match ch {
+            // period is illegal other than after a non-empty label
+            b'.' if label_length == 0 => {
+                return Err(InvalidDnsNameError);
+            }
+            // period is otherwise legal
+            b'.' => {
+                // but the last character of the previous label must not be a hyphen
+                if previous_label_ch == b'-' {
+                    return Err(InvalidDnsNameError);
+                }
+                label_length = 0;
+            }
+            // first character of label must be <letter>
+            b'a'..=b'z' | b'A'..=b'Z' if label_length == 0 => {
+                previous_label_ch = *ch;
+                label_length += 1;
+            }
+            // subsequent characters can be <letter> | <digit> | "-" | "_"
+            // (underscore is a relaxation of RFC1035)
+            b'a'..=b'z' | b'A'..=b'Z' | b'-' | b'_' | b'0'..=b'9' if label_length > 0 => {
+                previous_label_ch = *ch;
+                label_length += 1;
+            }
+            // other characters are illegal
+            _ => {
+                return Err(InvalidDnsNameError);
+            }
+        };
+
+        if label_length > MAX_LABEL_LENGTH {
+            return Err(InvalidDnsNameError);
+        }
+    }
+
+    if previous_label_ch == b'-' {
+        return Err(InvalidDnsNameError);
+    }
+
+    Ok(())
+}
+
+/// The provided input could not be parsed because
+/// it is not a syntactically-valid DNS Name.
+#[derive(Debug)]
+pub struct InvalidDnsNameError;
```

Comment:
Why not make this an enum with precise errors?

Reply:
I think making detailed/useful errors here makes things more complex and the benefit is relatively limited.